Allow onValidateContact plugin to show error message

beefcakefu opened the pull request:
Pull Request for Issue #37584 .

### Summary of Changes
This pull request copies actions performed on contact form submissions that fail native form validation, and applies it to submissions that fail `onValidateContact` plugins.



### Testing Instructions
This [dummy plugin](https://www.dropbox.com/s/v2x130bl10mbqsm/plg_contact_dummyvalidation.zip?dl=0) always fail contact form submissions and should demonstrate the issue:
```
<?php
defined('_JEXEC') or die;

class plgContactDummyValidation extends JPlugin {

	protected $autoloadLanguage = true;

	function onValidateContact($contact, $data) {
		return new InvalidArgumentException("Exception message is not shown.");
	}
}
```

1. Install dummy plugin above.
(System > Manage > Extensions > Install Extensions > Upload Package File)
2. Enable dummy plugin.
(System > Manage > Plugins > Contact - Dummy Validation > Enable plugin)
3. Create a contact form in admin panel.
(Menus > Main Menu > Add Site Menu Item > Menu Item Type > Contacts > Single Contact > Select Contact [create if there isn't already one], and give it a title)
4. Set up valid mail credentials if not already present.
(System > Setup > Global Configuration > Server > Mail)
5. Fill and submit contact form on the public site. Observe how user is redirected to a blank page.
6. Apply patch.
7. Fill and submit contact form again. Observe how user is redirected to contact form pre-filled with previously entered values, with exception message in an alert banner which plugins can use to explain why the submission was unsuccessful.

### Actual result BEFORE applying this Pull Request
If user submits a contact form with parameters failing an `onValidateContact` triggered plugin, plugin returns an exception, and user is redirected to a blank contact form page, stripped of all form elements with no error message indicating why the submission was unsuccessful.
![Screenshot from 2022-04-12 23-41-00](https://user-images.githubusercontent.com/103491197/163001047-f4673c35-2c1e-42e2-8a1e-6301bbdaca1a.png)



### Expected result AFTER applying this Pull Request
User is redirected to contact form page, pre-filled with submitted values and displaying exception message explanation of why submission failed. Same as if submission had failed native form validation.
![Screenshot from 2022-04-21 11-42-00](https://user-images.githubusercontent.com/103491197/164367612-a4833049-e77a-4506-b919-fbefda407634.png)



### Documentation Changes Required
None.

beefcakefu committed:
Allow onValidateContact listener to give feedback

If a plugin listening to the `onValidateContact` event wants to stop contact form submission, it is supposed to return an exception. Currently, after submission is halted, user is sent back to a blank contact form page, stripped of all form elements with no error message indicating why the submission was unsuccessful.

PR resolves this by copying the actions for submissions failing native form validation, which is to redirect to contact form pre-filled with submitted values and displaying exception message, hopefully a descriptive explanation of why submission failed.

I've written an email validation plugin that returns exceptions with descriptive messages that can be used to test out the PR. Download here: https://www.dropbox.com/s/s1m9j8uqq0689p9/plg_contact_cvalidation.zip?dl=0

beefcakefu committed:
Merge branch '4.1-dev' into patch-1

richard67 commented on components/com_contact/src/Controller/ContactController.php:
```diff
@@ -179,14 +179,26 @@ public function submit()
 			}
 		}
 
+		if(!$passValidation)
```
```suggestion
		if (!$passValidation)
```
Fix code style error.

beefcakefu committed:
Update components/com_contact/src/Controller/ContactController.php

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

beefcakefu committed:
Merge branch '4.1-dev' into patch-1

richard67 commented:
@beefcakefu I've noticed you have just used the "Update branch" button because the branch was shown as not up to date. That's not a problem. But it is also not necessary as long as there are not shown conflicting files in addition. So I just wanted to let you know it's not necessary to always do that.

Regarding your review request: Code style looks ok, but for a deeper review I don't have the time now.

PhilETaylor commented on components/com_contact/src/Controller/ContactController.php:
```diff
@@ -179,14 +179,26 @@ public function submit()
+		{
+			$app->setUserState('com_contact.contact.data', $data);
+
+			$this->setRedirect(Route::_('index.php?option=com_contact&view=contact&id=' . $stub . '&catid=' . $contact->catid, false));
```
@joomla/security `$stub` is a string at this point, thus possibility of XSS. Possibly use `$id` which is an integer instead?

beefcakefu replied:
Hey @PhilETaylor, thank you for the feedback! I've updated the reference to `$stub` with `$id` as suggested.

Additionally, I see that `$stub` was used similarly in 4 other places within the `submit()` function:

https://github.com/joomla/joomla-cms/blob/5a1d6c041df06692604a06dde68a268aa7834891/components/com_contact/src/Controller/ContactController.php#L123

https://github.com/joomla/joomla-cms/blob/5a1d6c041df06692604a06dde68a268aa7834891/components/com_contact/src/Controller/ContactController.php#L139

https://github.com/joomla/joomla-cms/blob/5a1d6c041df06692604a06dde68a268aa7834891/components/com_contact/src/Controller/ContactController.php#L174

https://github.com/joomla/joomla-cms/blob/5a1d6c041df06692604a06dde68a268aa7834891/components/com_contact/src/Controller/ContactController.php#L231

While `$id`, after declaration on line 86:

https://github.com/joomla/joomla-cms/blob/5a1d6c041df06692604a06dde68a268aa7834891/components/com_contact/src/Controller/ContactController.php#L86

Was only used once:

https://github.com/joomla/joomla-cms/blob/5a1d6c041df06692604a06dde68a268aa7834891/components/com_contact/src/Controller/ContactController.php#L93

I won't pretend to understand this potential XSS vulnerability anywhere near as well as you, so I'll leave you to decide if it's necessary to open another PR to fix these 4 lines above.

beefcakefu commented:
> @beefcakefu I've noticed you have just used the "Update branch" button because the branch was shown as not up to date. That's not a problem. But it is also not necessary as long as there are not shown conflicting files in addition. So I just wanted to let you know it's not necessary to always do that.
> 
> Regarding your review request: Code style looks ok, but for a deeper review I don't have the time now.

Thanks @richard67! As you know, I'm very new to this, and your guidance is much appreciated.

beefcakefu committed:
Changed $stub to $id to prevent XSS

Updating reference to $stub with $id at @PhilETaylor's suggestion to prevent XSS.

beefcakefu commented:
@richard67 Sorry to bug you again, but it looks like this PR has stalled. I hope I don't come off sounding demanding, because that's not my intention, and I totally understand everyone here is volunteering their free time for the Joomla! project.

I was hoping you could enlightening me about the typical lifecycle of a PR. At this point, I've addressed 2 reviews by yourself and @PhilETaylor. Does something need to happen from you guys, like to acknowledge that your feedbacks were satisfactorily resolved? Or do we just wait for 2 human testers to confirm the PR works and we're ready to commit?

I've actually authored a plugin to validate that enquirers' email addresses are working before allowing a contact form submission. It opens a connection to the enquirer's SMTP server and checks that the mailbox exists and isn't full and everything. I think it'll be a great plugin to combat prolific spammers like [eric.jones.z.mail@gmail.com](https://old.reddit.com/r/Wordpress/comments/na2xxc/form_spam_eric_jones/). I hope to release the plugin as open source, and the only thing stopping me from putting it on the Joomla! Extensions Directory is that exception messages currently cannot be shown to an enquirer without this PR.

This PR and my plugin also work in Joomla! 3.x. What do I have to do so this PR is also committed to the 3.x branch?

Thank you for your attention, and your patience when I opened the unnecessary Issue, and teaching me it's not necessary to always "update branch" everytime it shows not up to date.

PhilETaylor commented:
The Joomla project is well known for leaving PR contributions open as long as 5+ years… so 22 days open for this PR is nothing. If a PR is not tested and merged within 3 days then it normally stalls for months or even years. This is not opinion, this is fact. You can do no more except wait for someone to test your PR and for a maintainer to accept or reject it.

richard67 commented:
> Or do we just wait for 2 human testers to confirm the PR works and we're ready to commit?

@beefcakefu That's the case.

> What do I have to do so this PR is also committed to the 3.x branch?

You would have to make a PR for the 3.10-dev branch and mention this PR here in the description of the other PR so it's clear they do the same.

But let's ask the 3.10 release lead @zero-24 first if this is a bug fix which he would like to have in 3.10.

richard67 commented:
> I intend to start contributing as a patch tester and hopefully make a difference in this unclosed PR contribution situation.

@beefcakefu That would indeed be a good help because we are so few people doing that. Thanks in advance.

bembelimen committed:
Merge branch '4.2-dev' into patch-1

bembelimen commented:
Hey @beefcakefu very cool addition :+1: Thank you for your contribution.
I moved it to the 4.2 branch, as it's not really a bug fix but an enhancement.
So it just needs now two valid tests and we're good to go.

brianteeman commented:
> I moved it to the 4.2 branch, as it's not really a bug fix but an enhancement.

Which also means that it wont be accepted for j3

toivo commented:
Unable to report failed test result from Joomla 4.2.0-alpha3-dev of 6 May.  Is that caused by the out-of-date message "This branch is out-of-date with the base branch"?

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37540">issues.joomla.org/tracker/joomla-cms/37540</a>.</sub>

beefcakefu committed:
Merge branch '4.2-dev' into patch-1

beefcakefu commented:
> Unable to report failed test result from Joomla 4.2.0-alpha3-dev of 6 May. Is that caused by the out-of-date message "This branch is out-of-date with the base branch"?
> This comment was created with the [J!Tracker Application](https://github.com/joomla/jissues) at [issues.joomla.org/tracker/joomla-cms/37540](https://issues.joomla.org/tracker/joomla-cms/37540).

I think that was it! Have updated branch and it's now passing. Thanks toivo!

richard67 commented:
> Unable to report failed test result from Joomla 4.2.0-alpha3-dev of 6 May. Is that caused by the out-of-date message "This branch is out-of-date with the base branch"?

@beefcakefu What @toivo means is that he can't use the "Test this" button in the issue tracker here https://issues.joomla.org/tracker/joomla-cms/37540 to mark his test result.

@toivo That's a but which the issue tracker sometimes has. For some reason it did not see that this PR is a PR and not an issue, and so the status is "New" and not "Pending" in the issue tracker, and the button is missing. What you can do is you can report back your test result in a normal comment and provide some information about what has failed so @beefcakefu can fix his PR if necessary.

beefcakefu commented:
> @beefcakefu What @toivo means is that he can't use the "Test this" button in the issue tracker here https://issues.joomla.org/tracker/joomla-cms/37540 to mark his test result.
I see.

Thank you for taking the time to test this PR, @toivo. Where are you running into problems?

richard67 commented:
@Kubik-Rubik If you would scroll a bit up you would find my comment regarding the issue tracker problem.

joomla-bot committed:
Merge tag 'psr12anchor' into psr12/merge/37540

joomla-bot committed:
Phase 1 convert BRANCH to PSR-12

joomla-bot committed:
Phase 2 convert BRANCH to PSR-12

joomla-bot committed:
Merge tag 'psr12final' into psr12/merge/37540

HLeithner commented:
This pull request has been automatically rebased to 4.3-dev.

TLWebdesign commented:
Tested successfully on ‎4.3.5-dev<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37540">issues.joomla.org/tracker/joomla-cms/37540</a>.</sub>

RickR2H commented:
Tested successfully on ‎4.3.5-dev<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37540">issues.joomla.org/tracker/joomla-cms/37540</a>.</sub>

RickR2H commented:
RTC (Tests where added as comments as issue tracker had issues)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37540">issues.joomla.org/tracker/joomla-cms/37540</a>.</sub>

QuyTon committed:
Merge branch '4.3-dev' into patch-1

QuyTon committed:
Merge branch '4.3-dev' into patch-1

obuisard commented:
Thank you @beefcakefu for this PR.
I consider it a long lasting bug fix and took the responsibility to merge it.
